### PR TITLE
FE-1406 - [ Carousel Component ]

### PR DESCRIPTION
**Pull request by dawidzarzycki**
# Description
- Create new version of carousel component

# Related Issues / Pull Requests
- There is a className used by `Page Component`

# Screenshots / Gifs
- classic
Zrzut ekranu 2019-07-31 o 14 40 00

- DLS
Zrzut ekranu 2019-07-31 o 14 02 10

# Testing Instructions
- story of `classic` has only classic theme applied

### CODE REVIEW

File: change_log/next/FE-1406-carousel.yml [hunk 1]
```diff
@@ -0,0 +1 @@
+Component Update: "Add new DLS themes to the Carousel Component"
```
Comment (grabkowski) [inline change_log/next/FE-1406-carousel.yml]: I think that should be `Styling Changes: "Adds DLS styling..."`

File: src/components/carousel/carousel-classic.style.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  :hover{
+  background-color: #99ADB6;
+  color: #255BC7;
+  cursor: default;
+  }
+
+  &:focus {
+      outline: none;
+      border: none;
+  }
```
Comment (grabkowski) [inline src/components/carousel/carousel-classic.style.js]: could you fix the indentation?

File: src/components/carousel/carousel.style.js [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+  &&::before {
+    font-size: 16px;
+  }
+
+  ${({ theme }) => isClassic(theme) && css`
+    &&::before {
+      font-size: 25px;
+    } 
+  `}
```
Comment (grabkowski) [inline src/components/carousel/carousel.style.js]: I think that could be just:
```
&&::before {
  font-size: ${({ theme }) => (isClassic(theme) ? '25px' : '16px'});
}
```

File: src/components/carousel/slide.config.js [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+.carousel-transition-fade-leave {
+  opacity: 1;
+  position: relative;
+}
+
+.carousel-transition-fade-leave.carousel-transition-fade-leave-active {
+  opacity: 0;
+  transition: all 500ms cubic-bezier(.23,1.10,.69,1.01);
```
Comment (grabkowski) [inline src/components/carousel/slide.config.js]: couldn't this transition be a const?

File: src/components/carousel/slide/slide.style.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+      ${onClick && css`
+        :hover{
+          transition: all 0.2s ease-in;
+          transform: scale(1.02);
+          cursor: pointer;
+        }
+      `}
+    `}
+  `}
```
Comment (grabkowski) [inline src/components/carousel/slide/slide.style.js]: I think that could be extracted to a separate file

File: change_log/next/FE-1406-carousel.yml (hunk 1 above)
Comment (dawidzarzycki) [reply change_log/next/FE-1406-carousel.yml]: The whole component changed. Maybe should I write better description?

File: src/components/carousel/slide/slide.style.js (hunk 5 above)
Comment (dawidzarzycki) [reply src/components/carousel/slide/slide.style.js]: I thought about it but there was no paralel situation so I didn't know how to name the file.
Does the `slide-dls.js` name sounds good for you?

File: src/components/carousel/carousel.component.js [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+   * A lifecycle method that is called after before initial render.
+   * Can set up state of component without causing a re-render
+   */
+  componentWillMount() {
+    const selectedIndex = Number(this.props.slideIndex) || Number(this.props.initialSlideIndex);
+    this.setState({ selectedSlideIndex: selectedIndex });
+  }
+
+  /** A lifecycle method that is called before re-render. */
+  componentWillReceiveProps(nextProps) {
```
Comment (edleeks87) [inline src/components/carousel/carousel.component.js]: This is a legacy lifecycle method and needs updating to use `componentDidUpdate(prevProps)`

File: src/components/carousel/carousel.component.js [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+  enableButtonsAfterTimeout() {
+    if (isClassic(this.props.theme)) {
+      setTimeout(() => {
+        this.setState({ disabled: false });
+      }, TRANSITION_TIME);
+    }
+  }
+
+  /** Gets the props for the previous button */
+  previousButtonProps() {
```
Comment (edleeks87) [inline src/components/carousel/carousel.component.js]: As the prop object has only one internal property you could change this to a one liner:
```
return !this.state.disabled ? { onclick: this.onPreviousClick } : {};
```

File: src/components/carousel/carousel.component.js [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+
+    if (!this.state.disabled) {
+      props.onClick = this.onPreviousClick;
+    }
+
+    return props;
+  }
+
+  /** Gets the props for the next button */
+  nextButtonProps() {
```
Comment (edleeks87) [inline src/components/carousel/carousel.component.js]: Similar to my previous comment 
```
return !this.state.disabled ? { onclick: this.onNextClick } : {};
```

**Pull request merged**